### ISSUE TITLE
JavaAgent.transform not checking for null className

### ISSUE THREAD

**Issue by LesNovell** [puniverse/quasar#160]
I have run into this problem while developing a Vert.x Sync app which is based upon Quasar.   In JDK 8, it is now possible with Unsafe.defineAnonymousClass() that JavaAgent.transform() can receive a null className.

I suggest, in this case, that we skip attempts to transform this class.  Currently, I am seeing a NullPointerException on the first line of this method...:

```
        public byte[] transform(ClassLoader loader, String className, Class<?> classBeingRedefined, ProtectionDomain protectionDomain, byte[] classfileBuffer) throws IllegalClassFormatException {

            if (className.startsWith("clojure/lang/Compiler"))

```

For reference, Hibernate project has also run into this problem and proposed and implemented a null check solution. https://hibernate.atlassian.net/browse/HHH-9541

If you would like, I can fix and send a pull request.

Thanks,
Les Novell
Ticketmaster Sr. Software Engineer

**Comment by circlespainter** [comment]
Thanks for the insightful report.

Indeed there are few points even after that one where Quasar instrumentation assumes a class has some (possibly generated) name but my doubt is: is there a chance that methods in those anonymous classes need to be instrumented (which is the case if they can appear in runtime JVM stacks culminating with a fiber suspension)?

If this can happen, and I can't see why not if reasoning in general terms, then I think they should pass through instrumentation as well (I need to take a deeper look but after a quick one it might actually not be so hard to do it).

Do you have some pointer to / insight about an `Unsafe.defineAnonymousClass` usage in `vert.x`'s sync module (or elsewhere) that could require Quasar instrumentation? Or, if you think this can never be the case for classes defined through `Unsafe.defineAnonymousClass`, why is that so?

**Comment by LesNovell** [comment]
Hi Fabio, these are all great questions.  I have just run across this problem, I can't say I completely understand the conditions when Unsafe.defineAnonymousClass generates a null className.  This method looks like it's used by the JVM to create call-sites for lambdas.

It does seem very possible that a method in the unnamed class could require Quasar instrumentation, although this is not the case right now for me. But generally speaking, should the policy be to always attempt to instrument unnamed classes? 

Here's the stack trace details, as you can see this unnamed anonymous class occurs as a result of some Vert.x code.  I'm still not sure why Java can't give it a name... maybe it's a Java defect. I could not find it in the Java Bug Database.

```
NullPointerException at
	  at co.paralleluniverse.fibers.instrument.JavaAgent$Transformer.transform(JavaAgent.java:174)
	  at sun.instrument.TransformerManager.transform(TransformerManager.java:188)
	  at sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:428)
	  at sun.misc.Unsafe.defineAnonymousClass(Unsafe.java:-1)
	  at java.lang.invoke.InnerClassLambdaMetafactory.spinInnerClass(InnerClassLambdaMetafactory.java:324)
	  at java.lang.invoke.InnerClassLambdaMetafactory.buildCallSite(InnerClassLambdaMetafactory.java:194)
	  at java.lang.invoke.LambdaMetafactory.metafactory(LambdaMetafactory.java:304)
	  at java.lang.invoke.LambdaForm$DMH.1230013344.invokeStatic_L6_L(LambdaForm$DMH:-1)
	  at java.lang.invoke.LambdaForm$BMH.350413251.reinvoke(LambdaForm$BMH:-1)
	  at java.lang.invoke.LambdaForm$MH.1527906606.invoke_MT(LambdaForm$MH:-1)
	  at java.lang.invoke.CallSite.makeSite(CallSite.java:302)
	  at java.lang.invoke.MethodHandleNatives.linkCallSiteImpl(MethodHandleNatives.java:307)
	  at java.lang.invoke.MethodHandleNatives.linkCallSite(MethodHandleNatives.java:297)

--> Below is the line of Vert.x code that's causing Java to create the class without a name

	  at io.vertx.core.http.impl.HttpClientRequestImpl.connect(HttpClientRequestImpl.java:508)
	  - locked <0x1aeb> (a io.vertx.core.http.impl.HttpClientRequestImpl)
	  at io.vertx.core.http.impl.HttpClientRequestImpl.write(HttpClientRequestImpl.java:653)
	  at io.vertx.core.http.impl.HttpClientRequestImpl.write(HttpClientRequestImpl.java:182)
	  at io.vertx.core.http.impl.HttpClientRequestImpl.write(HttpClientRequestImpl.java:56)
	  at com.ticketmaster.vertx.feature.restclient.RestRequest.lambda$getHttpClientResponse$25(RestRequest.java:153)
	  at com.ticketmaster.vertx.feature.restclient.RestRequest$$Lambda$101.136075705.accept(Unknown Source:-1)
	  at io.vertx.ext.sync.Sync$2.requestAsync(Sync.java:66)
	  at co.paralleluniverse.fibers.FiberAsync$1.run(FiberAsync.java:127)
	  at co.paralleluniverse.fibers.Fiber.exec(Fiber.java:762)
	  at co.paralleluniverse.fibers.RunnableFiberTask.doExec(RunnableFiberTask.java:94)
	  at co.paralleluniverse.fibers.RunnableFiberTask.run(RunnableFiberTask.java:85)
	  at io.vertx.ext.sync.Sync.lambda$getContextScheduler$2(Sync.java:138)
	  at io.vertx.ext.sync.Sync$$Lambda$8.514941883.execute(Unknown Source:-1)
	  at co.paralleluniverse.fibers.FiberExecutorScheduler.execute(FiberExecutorScheduler.java:77)
	  at co.paralleluniverse.fibers.RunnableFiberTask.submit(RunnableFiberTask.java:265)
	  at co.paralleluniverse.fibers.Fiber.start(Fiber.java:1055)
	  at io.vertx.ext.sync.Sync.lambda$fiberHandler$0(Sync.java:88)
	  at io.vertx.ext.sync.Sync$$Lambda$55.1463832911.handle(Unknown Source:-1)
	  at io.vertx.ext.web.impl.RouteImpl.handleContext(RouteImpl.java:220)
	  - locked <0x131e> (a io.vertx.ext.web.impl.RouteImpl)
	  at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:78)
	  at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:93)
	  at com.ticketmaster.uapi.uwd.authentication.oauth.OAuth2ToBasicAuthConverter.handle(OAuth2ToBasicAuthConverter.java:72)
	  at com.ticketmaster.uapi.uwd.authentication.oauth.OAuth2ToBasicAuthConverter.handle(OAuth2ToBasicAuthConverter.java:27)
	  at io.vertx.ext.sync.Sync.lambda$null$19031fba$1(Sync.java:88)
	  at io.vertx.ext.sync.Sync$$Lambda$70.2080961621.run(Unknown Source:-1)
```

The Vert.x code at line 508 of HttpClientRequestImpl is the following.  It appears to be creating a lambda that is a implementation of io.vertx.code.Handler. Could this have something to do with the way Vert.x was compiled? I would have thought that this lambda would have been converted to an anonymous class at compilation time.
```
io.vertx.core.http.impl.HttpClientRequestImpl (line 508)

      client.getConnection(port, host, conn -> {
        synchronized (this) {
          if (exceptionOccurred) {

```

Let me know what you think,
Les Novell
Ticketmaster Sr. Software Engineer

**Comment by circlespainter** [comment]
Thanks for the additional information, I'll look into it more thoroughly.

> It does seem very possible that a method in the unnamed class could require Quasar instrumentation, although this is not the case right now for me. But generally speaking, should the policy be to always attempt to instrument unnamed classes?

Instrumentation will examine all loaded classes but will actually instrument only the ones that satisfy certain requirements, which at present are the ones that contain methods marked as suspendable in any of the supported ways (i.e. checked `throws SuspendExecution` clause, `@Suspendable` annotation, `META-INF/suspendables` / `META-INF/suspendable-supers` resources and registered `SuspendableClassifier`s) and that actually call into other methods identified as suspendable.

Assuming a class from `Unsafe.defineAnonymousClass` satisfies the requirements (suppose f.e. that some of its methods are annotated as `@Suspendable` and actually call into other suspendable methods) then in this case yes, I think it should be instrumented.

**Comment by LesNovell** [comment]
Ah... Ok, that makes sense. Yes, I guess it's very possible the functional interface to have throws SuspendExecution declared or have @Suspendable annotation.  Thanks for the clarification.  I look forward to a fix for this issue.

**Comment by circlespainter** [comment]
Closed through https://github.com/puniverse/quasar/pull/167. Thanks!

**Comment by LesNovell** [comment]
@circlespainter Hey I guess there in an unintended consequence of my change.  It seems like now the agent is trying to instrument anonymous classes that represent lambdas.  This is causing various ERRORs to be output, although there is no impact on the running code.  The lambda is not marked as Suspendable.  

What do you think a good solution to this might be?  These exceptions occur when first starting Vertx, so it's going to be a problem for production release.

Exceptions look like this...
``[quasar] ERROR: Unable to instrument class io/vertx/core/impl/OrderedExecutorFactory$OrderedExecutor
co.paralleluniverse.fibers.instrument.UnableToInstrumentException: Unable to instrument io/vertx/core/impl/OrderedExecutorFactory$OrderedExecutor#lambda$new$261()V because of synchronization
	at co.paralleluniverse.fibers.instrument.InstrumentMethod.dumpCodeBlock(InstrumentMethod.java:720)
	at co.paralleluniverse.fibers.instrument.InstrumentMethod.accept(InstrumentMethod.java:415)
	at co.paralleluniverse.fibers.instrument.InstrumentClass.visitEnd(InstrumentClass.java:265)
	at co.paralleluniverse.asm.ClassReader.accept(Unknown Source)
	at co.paralleluniverse.asm.ClassReader.accept(Unknown Source)
	at co.paralleluniverse.fibers.instrument.QuasarInstrumentor.instrumentClass(QuasarInstrumentor.java:120)
	at co.paralleluniverse.fibers.instrument.QuasarInstrumentor.instrumentClass(QuasarInstrumentor.java:88)
	at co.paralleluniverse.fibers.instrument.JavaAgent$Transformer.transform(JavaAgent.java:187)
	at sun.instrument.TransformerManager.transform(TransformerManager.java:188)
	at sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:428)
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:760)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at io.vertx.core.impl.OrderedExecutorFactory.getExecutor(OrderedExecutorFactory.java:53)
	at io.vertx.core.impl.VertxImpl.createEventLoopContext(VertxImpl.java:344)
	at io.vertx.core.impl.VertxImpl.getOrCreateContext(VertxImpl.java:310)
	at io.vertx.core.eventbus.impl.EventBusImpl.<init>(EventBusImpl.java:110)
	at io.vertx.core.impl.VertxImpl.<init>(VertxImpl.java:190)
	at io.vertx.core.impl.VertxImpl.<init>(VertxImpl.java:114)
	at io.vertx.core.impl.VertxFactoryImpl.vertx(VertxFactoryImpl.java:42)
	at io.vertx.core.Vertx.vertx(Vertx.java:88)
	at com.ticketmaster.helix.base.server.HelixServer$1.run(HelixServer.java:90)
	at java.lang.Thread.run(Thread.java:745)
``

**Comment by circlespainter** [comment]
I'm not completely sure it's related to #160, I think it's more likely caused by the fix for #168 (description of both issue and solution in PR #169).

Shortly: lambda literals will become `invokedynamic`s ultimately linked to method handles pointing to those (synthetic, generated at runtime) `lambda$`-prefixed methods. Now, there seems no way to annotate lambda literals as suspendable (see the PR for details) so they have to be assumed as such and this means the instrumentor will analyze them.
Code in lambda literals can use "old-style" Java built-in synchronization constructs (the new, more flexible and more efficient one being interface-based APIs in j.u.c of which Quasar provides a fiber-compatible implementation) which can only block threads and so by default the Quasar agent refuses to instrumented them (but it can be passed the `=m` option suffix to let them be instrumented anyway). As you noticed, these warnings are not fatal and they happen on `stderr` rather than `stdout` so they won't interfere with e.g. application logs, although they are surely annoying and can be even deceiving. But perhaps the complaining policy can be made softer, I'll have to look into and think about that a bit more.

I suggest to proceed as follows:

- I think `vert.x` should consider migrating to j.u.c locks rather than syncs and, since it supports fibers, it should also consider using the fiber-compatible implementation from Quasar. Could you talk about that on the group or perhaps even open an issue, still linking here?
- As it seems, this case you found is very likely a new problematic one originated from #169 so can you open a new issue and link to here for more info? We'll continue working there.
- In the meanwhile it should be enough to run the agent with the `=m` suffix (note that e.g. Pulsar's auto-instrumentation already requires that for very similar reasons).

**Comment by circlespainter** [comment]
@LesNovell OK, this is a pretty peculiar situation and I confirm the harmless but annoying error message has appeared due to #169 fixing #168.

That was a problem with synthetic methods implementing lambdas and making suspendable calls as they can't be annotated and it's not possible to tell during instrumentation which lambda interface they implement (in order to figure out if they should be considered suspendable or not); for this reason now they're always considered suspendable and analyzed.

Like any regular code, of course also lambdas can use `synchronized` and Quasar by default will trigger the error you see when trying to instrument a suspendable method containing it (because it will most like block threads). The error you see is thus caused by Quasar refusing to instrument this code https://github.com/eclipse/vert.x/blob/master/src/main/java/io/vertx/core/impl/OrderedExecutorFactory.java#L83.

The funny thing is that in this case it is harmless: that vert.x method is not really suspendable but it happens to access the private `log` field in the outer class, so it does that through a synthetic accessor method and accessors are conservatively marked "suspendable" by Quasar (lacking better local information about them, similarly to e.g. reflection).

Of course this only happens when using the agent, not AoT, and an easy workaround in this case is simply to run it with the `=m` option suffix (meaning "allow monitors").

Anyway I'm going to submit a PR to `vertx-core` and ask to widen the visibility of the `log` field to package-private so that this annoyance disappears in future release.
